### PR TITLE
CLIENTS: Send Audio Transcription

### LINKED ISSUE #212: CLIENTS: Transcript Audio File

### PR DESCRIPTION
**Pull request by Catalin-Andronie**
Closes #212

### CODE REVIEW

**Comment by hassanhabib** [comment]
@Catalin-Andronie post a screenshot of this client running please

File: Standard.AI.OpenAI/Models/Clients/AudioTranscriptions/Exceptions/ChatCompletionClientValidationException.cs [hunk 1]
```diff
@@ -0,0 +1,17 @@
+﻿// ---------------------------------------------------------------------------------- 
+// Copyright (c) The Standard Organization, a coalition of the Good-Hearted Engineers 
+// ----------------------------------------------------------------------------------
+
+using Xeptions;
+
+namespace Standard.AI.OpenAI.Models.Clients.AudioTranscriptions.Exceptions
+{
+    public class AudioTranscriptionClientValidationException : Xeption
```
Comment (hassanhabib) [inline Standard.AI.OpenAI/Models/Clients/AudioTranscriptions/Exceptions/ChatCompletionClientValidationException.cs:9]: Why's this file called `ChatCompletion ...`

**Comment by Catalin-Andronie** [comment]
> @Catalin-Andronie post a screenshot of this client running please

@hassanhabib Unfortunately this will not work as expected because of the next RESTFulSense feature requests:

1. https://github.com/hassanhabib/RESTFulSense/issues/132
2. https://github.com/hassanhabib/RESTFulSense/issues/130

----
1. Without https://github.com/hassanhabib/RESTFulSense/issues/132 implemented we are forced to mark all `AudioTranscriptionRequest` properties as `string`. E.g: `Temperature` will become a `string` type instead of a `double` as it should be.

```diff
public class AudioTranscriptionRequest
{
-    public double Temperature { get; set; } = 0.2;
+    public string Temperature { get; set; } = "0.2";
}
```

2. Without https://github.com/hassanhabib/RESTFulSense/issues/130 implemented we are forced to set all the `AudioTranscriptionRequest` properties with some value even if some of them are optional. For the creation of a translation, OpenAI requires only the file and model to be present in the request body and the rest of the options are optional, and if we do so we are getting an exception for the other properties since they are null or have empty values.

```diff
var inputAudioTranscription = new AudioTranscription
{
    Request = new AudioTranscriptionRequest
    {
        Content = fileContent,
        FileName = fileName,
        Model = "whisper-1",
-       Prompt = null, // This is an optional value and I should not be forced to set its value.
+       Prompt = "Some prompt...",
-       Language = "", // This will throw an exception since RESTFulSense doesn't accept empty values.
+       Language = "en"
    }
};

AudioTranscription responseAudioTranscription =
    await this.openAIClient.AudioTranscriptions.SendAudioTranscriptionAsync(
        inputAudioTranscription);
```

@hassanhabib what's your suggestion?

File: Standard.AI.OpenAI.Tests.Unit/Services/Foundations/AudioTranscriptions/AudioTranscriptionServiceTests.cs [hunk 2]
```diff
@@ -71,8 +71,8 @@ private static dynamic CreateRandomAudioTranscriptionProperties()
                         .AreEqual;
         }
 
-        private static decimal CreateRandomDecimal()
-            => new SequenceGeneratorDecimal().GetValue();
+        private static double CreateRandomDouble()
+            => new SequenceGeneratorDouble().GetValue();
```
Comment (hassanhabib) [inline Standard.AI.OpenAI.Tests.Unit/Services/Foundations/AudioTranscriptions/AudioTranscriptionServiceTests.cs:75]:
```suggestion
        private static double CreateRandomDouble() => 
            new SequenceGeneratorDouble().GetValue();
```

**Comment by BrianLParker** [comment]
Reading the API the file in the request is a string. The file should be previously uploaded using the name (string) for this. RESTFulSense already supports this using the `PostContentAsync` method. If you look at the example on the API page they are only posting the name of the file "german.m4a".  Sorry for the late response been quite ill for the last 4 days.

**Comment by Catalin-Andronie** [comment]
> [Reading the API](https://platform.openai.com/docs/api-reference/audio/create) the file in the request is a string. The file should be previously uploaded using the name (string) for this.

[The API](https://platform.openai.com/docs/api-reference/audio/create) only accepts `multipart/form-data` content-type, which means we cannot use `PostContentAsync` since that uses "json" content-type (See example request).
```curl
curl https://api.openai.com/v1/audio/transcriptions \
  -H "Authorization: Bearer $OPENAI_API_KEY" \
  -H "Content-Type: multipart/form-data" \
  -F file="@/path/to/file/audio.mp3" \
  -F model="whisper-1"
```

> RESTFulSense already supports this using the `PostContentAsync` method. If you look at the example on the API page they are only posting the name of the file "german.m4a". Sorry for the late response been quite ill for the last 4 days.

Actually, the API specifies to post the file itself as a stream of bytes, the example only illustrates what it should be there. Also, besides the `file` and `model` the API accepts other properties in the request body like `temperature`, `language`, and `prompt` (the latter are optional). This is the full model that can be posted:

Currently, we are using `PostFormAsync` in order to send the file stream to the server in order to obtain its transcription, which works like a charm. Even so, we have two issues/features which need to be implemented before moving forward and completing the Audio Transcription. Please [read this comment](https://github.com/hassanhabib/Standard.AI.OpenAI/pull/381#issuecomment-1546596171) to understand the RESTFulSense features.

**Pull request closed without merging**